daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +796,41 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+case class DFDLXLeftShift(recipes: List[CompiledDPath],argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift=shiftLong.toInt
+    argType match {
+      case NodeInfo.Long => {
+        if (shift >= 64) 0 else arg1.getLong << shift
+      }
+      case NodeInfo.Int => {
+        if (shift >=32) 0 else arg1.getInt<< shift
+      }
+      case NodeInfo.Short => {
+        if (shift >=16) 0 else DataValue.toDataValue((arg1.getShort << shift).toShort)
+      }
+      case NodeInfo.Byte => {
+        if (shift >=8) 0 else DataValue.toDataValue((arg1.getByte << shift).toByte)
+      }
+      case NodeInfo.UnsignedLong=>{
+        if(shift >=64) 0 else arg1.getLong << shift
```

Comment:
So,  since both Long and UnsignedLong are the same, scala lets you do this:
```
case NodeInfo.Long | NodeInfo.UnsignedLong =>  if (shift >= 64) 0 else arg1.getLong << shift
```
You may be able to double up all the signed and unsigned types that way, because for left shift, the code doesn't care whether it is signed or unsigned.

Reply:
Ah. I take it back, only Long and UnsignedLong are the same, because the others have the difference in size where an unsigned one is represented by the next larger integer.

Reply:
I don't think and signed types are the same as the unsigned types. In order to fit the full range of unsigned type, the underlying java primitive is one primitive bigger. So for NodeInfo.Long, arg1 is a java Long, but for NodeInfo.UnsignedLong, arg1 is a BigInt. As another example NodeInfo.Int is a java Int, but NodeInfo.UnsignedInt is a java Long. The get code to getLong vs getInt vs getBigInt, etc. is all going to be different.

We probably needs tests for every one of these primitive types to ensure they all work. Would be nice to also have multiple edge cases coverged, e.g. 0xFFFFFFFFF shift left 1, shift left 32, etc..

Reply:
Also  I noticed left shifting an int by 32 or long by 64 and so on returns the same number.(for my scala compiler)
Can you confirm do we have to return 0 for them or the number?

Reply:
I can comfirm this behavior. And after some googling, that's actually the expected behavior in Scala/Java:

https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.19

So if the shifted value is an int, only the 5 low order bits of the shift value are actually used--it's 6 for a long. So a shift left of an int of N bis is actually more like a shift left of ``N % 32``. So this is the importance of the >= 32, 64, etc. check to always return zero.

Alternatively, we might be to just define dfdlx:shiftLeft behavior to match Java's behavior. From what I can tell, in C a shift left greater than the width of the type is undefined. Java's behavior feels unintuitive to me, but at least it's defined...

Reply:
I would never guess that an int shift by 32 or greater is a no-op. I fully expect it to be zero, because you have shifted the bits out of the word. If it's going to have weird behavior on values out of range, it should throw an exception. To me, the only sensible definition for shift of word-size or greater is either 0, or throw.

For logical shift right, the same is true.

For arithmetic shift right, if the value is signed and negative, then arithmetic shift right by word-size or greater is -1. 

Again, or it could throw an argument range exception. In our case cause a processing error.

Reply:
Yeah, 0 (or -1 for shift right) or an SDE seems like the right behavior. Must be a reason for Java's behavior, but it seems very counter intuitive.

I kindof like SDE. If you're shifting by something larger than the width of a type, I imagine it's an unintentional error in the vast majority of cases. And if a user really wants the zero behavior, they can always wrap it in a similar condition we have and return zero instead of doing the shift.

Reply:
I think this would be a processing error, not a schema definition error. Dividing by zero is a processing error, and array access out of bounds is a processing error, so I think this would be as well. 

An SDE is more conservative in the sense of preserving the future freedom to change our minds about this. A processing error can be depended upon to force backtracking, so once it's a processing error, it's incompatible to take it back and make the behavior legal. 

So I guess I will throw my vote behind making it a Schema Definition Error.